### ISSUE TITLE
Problem with upload on edit with save Potlatch v1.0 (post Changeset 15276)

### ISSUE THREAD

**Issue by tomhughes** [tomhughes/trac-tickets#1904]
**[Submitted to the original trac issue database at 12.01pm, Tuesday, 2nd June 2009]**

Over the last couple of days (actually my first edits since 25th May) I've noticed problems with Potlatch saves failing.  What seems to happen is this:

o Part way through the save (typically in the changes to ways) Potlatch indicates that the save failed.

o There's a "retry" button, but pressing it after a few seconds, a few minutes or an hour doesn't work - it immediately gives a message such as "upload failed"*.

o Another Potlatch session will have no problem saving at this time - it's as if the state that the original session is in at the "retry" error is "confused" - it can't apply changes because some have already been committed.

What would be nice would be if Potlatch's saves could be better integrated with database commits - currently it seems to work independently of them.  If a save error occurs, would it be possible for Potlatch to abandon whatever it has "ready to commit" in a changeset and apply changes from current database state?  This would at least avoid the problem where someone edits a way into 2, and the "truncate" happens but the "add new way" doesn't.

For an example of a changeset that results from a failed save see:
http://www.openstreetmap.org/browse/changeset/1401640

That has 3 nodes and 1 way in it.  I then redid the edits manually resulting in:

http://www.openstreetmap.org/browse/changeset/1401651
which has 39 nodes and 1 way.

and
http://www.openstreetmap.org/browse/changeset/1401695
which has 24 nodes and 6 ways.

I'm logging this as a new ticket because I've only seen problems since http://trac.openstreetmap.org/changeset/15276 went in.  This may be co-incidence, but the continuing issues suggest that there's still an issue with retry handling.

* sorry I don't have more details - users always are crap at reporting error messages aren't they?

**Comment by tomhughes** [comment]
**Author: Richard**
**[Added to the original trac issue at 1.18pm, Tuesday, 2nd June 2009]**

Hm. If it **immediately** gives an error, that sounds like it's not even touching the server - maybe Flash Player has closed the RemoteConnection. (Which browser/Flash Player version are you using?)

I wonder if forcing it to drop the connection, then reconnect, on 'retry' might fix this.

**Comment by tomhughes** [comment]
**Author: Richard**
**[Added to the original trac issue at 1.19pm, Tuesday, 2nd June 2009]**

(Incidentally I'm not quite sure what you mean in the fifth para - in 'edit with save' mode, each write is contingent on the one before executing correctly anyway.)

**Comment by tomhughes** [comment]
**Author: Richard**
**[Added to the original trac issue at 1.36pm, Wednesday, 3rd June 2009]**

#1907 is a duplicate of this.

**Comment by tomhughes** [comment]
**Author: SomeoneElse**
**[Added to the original trac issue at 9.53pm, Thursday, 4th June 2009]**

The first time that it fails, it's not immediately giving an error, it's part way through the upload (after some but not all of the ways have been uploaded).  Subsequent times that it fails it does seem to fail "immediately".

In the fifth para I'm assuming that Potlatch saves are completely independent of database commits - if an upload fails it's got an incomplete changeset to apply.  I was wondering if tying Potlatch saves to database commits would allow "retry save" to work easier (as the current database state would always be what Potlatch is applying changes to, not current database state plus some previously updated changes).  Obviously this ignores someone else editing in the same area (and if it's relevant I've had Potlatch report "another personhas modified that way since you did" when the "another person" was me, but it doesn't seem related to this problem).

The Browser is Firefox 3.0.10 on Windows XP (although I note that the person who logged #1907 saw it on a Mac).  I suspect that the relevant plugin is "Shockwave Flash 9.0 r115".  "Noscript" is installed but shouldn't be blocking anything relevant to Flash.  There's also a software firewall installed but again that shouldn't be blocking anything relevant.

It's also worth mentioning that problems have occurred when the site is "quite busy", but as it's "quite busy" most of the time that's not too helpful...

**Comment by tomhughes** [comment]
**Author: SomeoneElse**
**[Added to the original trac issue at 11.45pm, Monday, 8th June 2009]**

Yes - seems fixed to me.  Thanks!
